docs/config.md
# Configuration

## Introduction

When starting the program a config file `config.json` is passed to the application. It defines the core execution
parameters.

## Attributes and values

The following table lists all attributes and their possible values.

| Key                       | optional | Description                                                                                                                                           | Min | Max       | Default Value | Example Value                                    |
|---------------------------|----------|-------------------------------------------------------------------------------------------------------------------------------------------------------|-----|-----------|---------------|--------------------------------------------------|
| no_rounds                 | no       | Number of rounds to run the simulation for                                                                                                            | 1   | -         | -             | 10                                               |
| algorithm                 | no       | List of algorithms to simulate                                                                                                                        | -   | -         | -             | `["PowD", "FedCS", "random", "ActiveFL", "CEP"]` |
| alogrithm_config          | yes      | Algorithm configuration                                                                                                                               | -   | -         | -             | See algorithm documentation                      |
| no_epochs                 | yes      | Number of epochs to run on each client per round                                                                                                      | 1   | -         | 1             | 1                                                |
| no_clients                | no       | Number of clients to simulate                                                                                                                         | 1   | -         | -             | 1000                                             |
| batch_size                | yes      | Batch size, will affect behavior of training and validation. Can lead to errors with batchnorm, etc. if changed                                       | 1   | -         | 32            | 32                                               |
| verbose                   | yes      | Enables additional logging                                                                                                                            | -   | -         | `true`        | `true` or `false`                                |
| device                    | yes      | Pytorch device string                                                                                                                                 | -   | -         | `cpu`         | `cpu` or `cuda`                                  |
| num_cpu_per_client        | yes      | Number of CPU cores to assign to each client. Integer.                                                                                                | 1   | CPU_COUNT | 2             | 2                                                |
| num_gpu_per_client        | yes      | Number of GPU resources to assign to each client. Only used if `device` is set to `cuda`. Partial allocation is possible.                             | 0   | -         | 0.1           | 0.1                                              |
| timeout                   | no       | Timeout imposed on clients, in seconds. This affects only the simulation.                                                                             | 1   | -         | -             | 60                                               |
| generate_client_states    | yes      | Whether to generate new client states when running the simulation. <br/>If `False` is selected, errors may occur if not enough states are predefined. | -   | -         | `true`        | `true` or `false`                                |
| client_state_file         | no       | File to use for client states. Will be overwritten if generate_client_states is `True`.                                                               | -   | -         | -             | `client_states.csv`                              |
| distribute_data           | yes      | Whether to regenerate the data distribution.                                                                                                          | -   | -         | `true`        | `true` or `false`                                |
| data_distribution_file    | no       | File containing an existing data distribution, not used when distribute_data is  `True`                                                               | -   | -         | -             | `data_distribution.csv`                          |
| output_dir                | no       | Folder prefix to use for output. Will not be overwritten.                                                                                             | -   | -         | -             | `output`                                         |
| client_configuration_file | no       | File that contains the available client configurations. See documentation on Client Configuration.                                                    | -   | -         | -             | `client_configurations.csv`                      |
| max_workers               | yes      | Maximum number of workers that should be used when running distributed tasks. (May not always be enforced)                                            | -   | -         | 32            | 32                                               |
| data_config               | yes      | Configuration for data skew, distortion                                                                                                               | -   | -         | -             | See data configuration                           |
| validation_config         | yes      | Configuration of the final validation runs                                                                                                            | -   | -         | -             | See validation configuration                     |
| simulation_config         | yes      | Configuration relating to performance, reliability and network bandwidth of clients                                                                   | -   | -         | -             | See simulation configuration                     |
| base_strategy             | yes      | Base strategy to use for aggregation                                                                                                                  | -   | -         | `FedAvg`      | `FedAvg`, `FedAvgM`, `FedMedian`                 |


docs/simulation/configuration.md
The simulation can be configured in the ```config.json``` file. The following parameters can be set:

| Key                         | optional | Description                                                             | Min | Max | Default Value        | Example Value |
|-----------------------------|----------|-------------------------------------------------------------------------|-----|-----|----------------------|---------------|
| network_bandwidth_mean      | yes      | Mean network bandwidth to assign to clients                             | -   | -   | 20.0                 | 10            |
| network_bandwidth_std       | yes      | Standard deviation of client network bandwidth                          | -   | -   | 10.0                 | 10            |
| network_bandwidth_min       | yes      | Minimum network bandwidth                                               | 0.0 | -   | 0.0                  | 0             |
| performance_factor_mean     | yes      | Mean performance factor (execution time multiplier)                     | -   | -   | 1.0                  | 1             |
| performance_factor_std      | yes      | Standard deviation of the performance factor                            | -   | -   | 0.2                  | 0.2           |
| reliability_parameter       | yes      | Parameter for the exponential distribution of reliability               | -   | -   | 10.0                 | 10.0          |
| number_of_performance_tiers | yes      | Number of performance tiers from the `client_configurations.csv` to use | 1   | -   | 4                    | 4             |
| state_simulation_seed       | yes      | Seed to use in state simulation                                         | 0   | -   | 12367123871238713871 | -             |

seleceval/util/config.py
"""
This module contains the Config class which is used to parse the configuration file and validate the parameters.
"""
import json
import os

from cerberus import Validator
from datetime import datetime
from .config_parameters import *

from .config_parameters.feature_distribution_parameters import (
    data_feature_distribution_parameters,
    data_feature_distributions,
    default_feature_distribution,
)


class Config:
    """
    This class is used to parse the configuration file and validate the parameters.
    """

    def __init__(self, file_name: str):
        schema = {
            "no_rounds": {"type": "integer", "min": 1},
            "algorithm": {"type": "list", "allowed": algorithm_parameter_dict.keys()},
            "dataset": {"type": "string", "allowed": ["cifar10", "mnist"]},
            "algorithm_config": {
                "type": "dict",
                "default": {},
                "schema": algorithm_parameter_dict,
            },
            "no_epochs": {"type": "integer", "min": 1, "default": 1},
            "no_clients": {"type": "integer", "min": 1},
            "batch_size": {"type": "integer", "min": 1, "default": 32},
            "validation_split": {"type": "float", "min": 0, "max": 1, "default": 0.1},
            "device": {"type": "string", "allowed": ["cuda", "cpu"], "default": "cpu"},
            "num_cpu_per_client": {"type": "integer", "min": 1, "max": os.cpu_count(), "default": 2},
            "num_gpu_per_client": {"type": "float", "min": 0, "default": 0.1},
            "verbose": {"type": "boolean", "default": True},
            "timeout": {"type": "integer", "min": 1},
            "generate_clients": {"type": "boolean", "default": True},
            "client_state_file": {"type": "string"},
            "distribute_data": {"type": "boolean", "default": True},
            "data_distribution_file": {"type": "string"},
            "output_dir": {"type": "string"},
            "client_configuration_file": {"type": "string"},
            "validation_config": {
                "type": "dict",
                "default": {},
                "schema": {
                    "enable_validation": {"type": "boolean", "default": True},
                    "enable_data_distribution": {"type": "boolean", "default": True},
                    "device": {
                        "type": "string",
                        "allowed": ["cuda", "cpu"],
                        "default": "cpu",
                    },
                },
            },
            "simulation_config": {
                "type": "dict",
                "default": {},
                "schema": {
                    "network_bandwidth_mean": {"type": "float", "default": 20.0},
                    "network_bandwidth_std": {"type": "float", "default": 10.0},
                    "network_bandwidth_min": {
                        "type": "float",
                        "default": 0.0,
                        "min": 0.0,
                    },
                    "performance_factor_mean": {"type": "float", "default": 1.0},
                    "performance_factor_std": {"type": "float", "default": 0.2},
                    "reliability_parameter": {"type": "float", "default": 10.0},
                    "number_of_performance_tiers": {
                        "type": "integer",
                        "min": 1,
                        "default": 4,
                    },
                    "state_simulation_seed": {
                        "type": "integer",
                        "min": 0,
                        "default": 12367123871238713871,
                    },
                },
            },
            "max_workers": {"type": "integer", "min": 1, "default": 32},
            "base_strategy": {
                "type": "string",
                "allowed": available_strategies,
                "default": default_strategy,
            },
            "data_config": {
                "type": "dict",
                "default": {},
                "schema": {
                    "data_label_distribution_skew": {
                        "type": "string",
                        "allowed": data_label_distributions,
                        "default": default_label_distribution,
                    },
                    "data_quantity_skew": {
                        "type": "string",
                        "allowed": data_quantity_distributions,
                        "default": default_quantity_distribution,
                    },
                    "data_feature_skew": {
                        "type": "string",
                        "allowed": data_feature_distributions,
                        "default": default_feature_distribution,
                    },
                    **data_label_distribution_parameters,
                    **data_quantity_distribution_parameters,
                    **data_feature_distribution_parameters,
                },
            },
        }

        # Set data_config_schema

        v = Validator(schema, require_all=True)
        self.current_round = 0
        with open(file_name) as json_file:
            config_dict = json.load(json_file)
            if not v.validate(config_dict):
                raise ValueError(v.errors)
            self.initial_config = v.normalized(config_dict)

        self.initial_config["output_dir"] = (
            self.initial_config["output_dir"]
            + "_"
            + datetime.now().strftime("%Y%m%d_%H%M%S")
        )
        self.attributes = {
            "input_state_file": self.initial_config["output_dir"] + "/input_state.csv",
            "working_state_file": self.initial_config["output_dir"]
            + "/working_state.csv",
            "data_distribution_output": self.initial_config["output_dir"]
            + "/data_distribution.csv",
        }
        # If necessary create output dir + subdirs
        if not os.path.isdir(self.initial_config["output_dir"]):
            os.mkdir(path=self.initial_config["output_dir"])
        if not os.path.isdir(self.initial_config["output_dir"] + "/client_output"):
            os.mkdir(path=self.initial_config["output_dir"] + "/client_output")
        if not os.path.isdir(self.initial_config["output_dir"] + "/model_output"):
            os.mkdir(path=self.initial_config["output_dir"] + "/model_output")
        if not os.path.isdir(self.initial_config["output_dir"] + "/validation"):
            os.mkdir(path=self.initial_config["output_dir"] + "/validation")
        if not os.path.isdir(self.initial_config["output_dir"] + "/state"):
            os.mkdir(path=self.initial_config["output_dir"] + "/state")
        if not os.path.isdir(self.initial_config["output_dir"] + "/data_distribution"):
            os.mkdir(path=self.initial_config["output_dir"] + "/data_distribution")
        if not os.path.isdir(self.initial_config["output_dir"] + "/figures"):
            os.mkdir(path=self.initial_config["output_dir"] + "/figures")

    def set_current_round(self, i: int):
        """
        Sets the current round
        :param i: The current round
        :return: None
        """
        self.current_round = i

    def get_current_round(self):
        """
        Returns the current round
        :return: The current round
        """
        return self.current_round

    def generate_paths(self, algorithm: str, dataset: str, no_clients: int):
        """
        Generates the paths for the output files
        :param algorithm: Current algorithm simulated
        :param dataset: Current dataset used
        :param no_clients: Number of clients used
        :return: None
        """
        self.attributes["output_path"] = (
            self.initial_config["output_dir"]
            + "/client_output/"
            + "client_output_"
            + algorithm
            + "_"
            + dataset
            + "_"
            + str(no_clients)
            + ".json"
        )
        self.attributes["model_output_prefix"] = (
            self.initial_config["output_dir"]
            + "/model_output/"
            + "model_output_"
            + algorithm
            + "_"
            + dataset
            + "_"
            + str(no_clients)
            + "_"
        )
        self.attributes["state_output_prefix"] = (
            self.initial_config["output_dir"]
            + "/state/"
            + "state_"
            + algorithm
            + "_"
            + dataset
            + "_"
            + str(no_clients)
            + "_"
        )
